### PR TITLE
fix(datastore): decode optimistically with paginated list response type

### PR DESCRIPTION
**Pull request by 5d**
## Issue \#

https://github.com/aws-amplify/amplify-swift/issues/3259

## Description

- decoding the `PaginatedList` optimistically

## General Checklist

- [ ] Added new tests to cover change, if needed
- [x] Build succeeds with all target using Swift Package Manager
- [x] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [x] PR title conforms to conventional commit style
- [ ] New or updated tests include `Given When Then` inline code documentation and are named accordingly `testThing_condition_expectation()`
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift [hunk 1]
```diff
@@ -94,4 +95,46 @@ class PaginatedListTests: XCTestCase {
         }
     }
 
+    func testDecodePaginatedListOptimastically() {
```
Comment (thisisabhash) [inline AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift:98]: minor nit: ...Optimistically...

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift [hunk 2]
```diff
@@ -94,4 +95,46 @@ class PaginatedListTests: XCTestCase {
...
+        let syncQueryJSON = """
```
Comment (thisisabhash) [inline AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift:99]: we can declare this JSON along with other JSONs at the top and name it something like `syncQueryJSONWithMissingFields`

Comment (thisisabhash) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift:99]: we can also have a test with null object in `items` array: 
```
"items" : [
   null,
   { post }
]
```

**Review by thisisabhash — CHANGES_REQUESTED** [review]

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift (hunk 1 above)
Comment (5d) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift]: 👍 will update it

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift (hunk 2 above)
Comment (5d) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift:99]: will add `null` as a case

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift [hunk 3]
```diff
@@ -94,4 +95,47 @@ class PaginatedListTests: XCTestCase {
         }
     }
 
+    func testDecodePaginatedListOptimistically() {
```
Comment (thisisabhash) [inline AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift:98]: minor nit: given when then statement

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift (hunk 2 above)
Comment (thisisabhash) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift:99]: thanks for the updates, we should move the declaration of this new JSON string at the top of the file like at line 42

Comment (5d) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift:99]: This is not shared across test cases. Including test data inside the test case is easier to read.

File: AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift (hunk 3 above)
Comment (5d) [reply AmplifyPlugins/Core/AWSPluginsCoreTests/Sync/PaginatedListTests.swift:104]: updated the comment

**Review by thisisabhash — APPROVED** [review]
